File: packages/shared/src/schemas.ts
import { z } from 'zod';

export const riskBucketSchema = z.enum(['LOW', 'MED', 'HIGH']);

export const driveSchema = z.object({
  driveId: z.string(),
  model: z.string(),
  datacenter: z.string(),
  capacityBytes: z.number().int().nonnegative(),
  firstSeen: z.string(),
  lastSeen: z.string(),
});

export const featurePayloadSchema = z.object({
  drive_id: z.string(),
  day: z.string(),
  features: z.record(z.number().nullable()),
});

export const scoreResponseSchema = z.object({
  drive_id: z.string(),
  day: z.string(),
  risk_score: z.number().min(0).max(1),
  risk_bucket: riskBucketSchema,
  top_reasons: z.array(z.object({
    code: z.string(),
    contribution: z.number(),
    direction: z.enum(['UP', 'DOWN']),
  })),
  model_version: z.string(),
  scored_at: z.string(),
});

export type RiskBucket = z.infer<typeof riskBucketSchema>;
export type Drive = z.infer<typeof driveSchema>;
export type FeaturePayload = z.infer<typeof featurePayloadSchema>;
export type ScoreResponse = z.infer<typeof scoreResponseSchema>;

File: services/model/app/schemas.py
from __future__ import annotations

from datetime import date, datetime
from enum import Enum
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field


class SchemaModel(BaseModel):
    model_config = ConfigDict(protected_namespaces=(), extra="forbid", allow_inf_nan=False)


class RiskBucket(str, Enum):
    LOW = "LOW"
    MED = "MED"
    HIGH = "HIGH"


class ScoreRequest(SchemaModel):
    drive_id: str = Field(min_length=1, max_length=128, pattern=r"^\S+$")
    day: date
    features: dict[Annotated[str, Field(min_length=1, max_length=128)], Annotated[float, Field(strict=True, allow_inf_nan=False)] | None] = Field(min_length=1, max_length=256)


class ReasonCode(SchemaModel):
    code: str
    contribution: float
    direction: Literal["UP", "DOWN"]


class ScoreResponse(SchemaModel):
    drive_id: str
    day: date
    risk_score: float = Field(ge=0.0, le=1.0)
    risk_bucket: RiskBucket
    top_reasons: list[ReasonCode]
    model_version: str
    scored_at: datetime


class BatchScoreRequest(SchemaModel):
    items: list[ScoreRequest] = Field(min_length=1, max_length=1000)


class HealthResponse(SchemaModel):
    status: str
    model_loaded: bool
    model_version: str


class ModelInfoResponse(SchemaModel):
    model_version: str
    model_type: str
    horizon_days: int
    features: list[str]
    metrics: dict[str, float | list[dict[str, float]] | str]
    provenance: dict = Field(default_factory=dict)
    explanation_method: str = "unavailable"
